src/Hl7.Fhir.Support.Poco.Tests/Model/DateTests.cs

```diff
@@ -0,0 +1,152 @@
+        public void UpdatesCachedValue()
+        {
+            var dft = new Date(2023, 07, 11);
+            dft.TryToDateTimeOffset(out var dto).Should().BeTrue();
+            dft.TryToDateTimeOffset(out var dto2).Should().BeTrue();
+            dto.Equals(dto2).Should().BeTrue();
+
+            dft.Value = "2023-07-11";
+            dft.TryToDateTimeOffset(out dto).Should().BeTrue();
+            dto.Day.Should().Be(11);
+            dft.TryToDateTimeOffset(out dto2).Should().BeTrue();
+            dto.Equals(dto2).Should().BeTrue();
+
+            dft.ObjectValue = "2023-07-11";
+            dft.TryToDateTimeOffset(out dto).Should().BeTrue();
+            dto.Month.Should().Be(7);
+            dft.TryToDateTimeOffset(out dto2).Should().BeTrue();
+            dto.Equals(dto2).Should().BeTrue();
+
+            dft.Value = null;
+            dft.TryToDateTimeOffset(out _).Should().BeFalse();
+            Assert.ThrowsException<InvalidOperationException>(() => dft.ToDateTimeOffset());
+        }
+
+        [TestMethod]
+        public void ToDateTimeOffsetThrowsInvalidFormat()
+        {
+            var dft = new Date("T45:45:56");
+
+            Assert.ThrowsException<FormatException>(() => dft.ToDateTimeOffset());
+
+            dft.TryToDateTimeOffset(out var _).Should().BeFalse();
+        }
+
+        [TestMethod]
+        public void CanConvertToDateTime()
+        {
+            var dft = new Date(2023, 07, 11);
+            dft.TryToDate(out var dt).Should().BeTrue();
+            dt.Days.Should().Be(11);
+            dt.Precision.Should().Be(ElementModel.Types.DateTimePrecision.Day);
+
+            dft = new Date(2023, 7);
+            dft.TryToDate(out dt).Should().BeTrue();
+            dt.Days.Should().BeNull();
+            dt.Precision.Should().Be(ElementModel.Types.DateTimePrecision.Month);
+
+            dft = new Date(null);
+            dft.TryToDate(out dt).Should().BeTrue();
+            dt.Should().BeNull();
+        }
+
+        [TestMethod]
+        public void CacheImprovesSpeed()
+        {
+            var dts = "2023-07-11";
+            var dt = new Date(dts);
+            _ = dt.TryToDate(out var _); // trigger initial compile of regex
+
+            var sw = Stopwatch.StartNew();
+
+            for (var i = 0; i < 1000; i++)
+            {
+                // Clear the cache each invocation
+                dt.Value = dts;
+                _ = dt.TryToDate(out var _);
+                dt.Value = null;
+            }
+
+            sw.Stop();
+            Console.WriteLine(sw.Elapsed.ToString());
+
+            dt = new Date(dts);
+
+            var sw2 = Stopwatch.StartNew();
+            for (var i = 0; i < 1000; i++)
+            {
+                _ = dt.TryToDate(out var _);
+            }
+            sw2.Stop();
+
+            Console.WriteLine(sw2.Elapsed.ToString());
+
+            // It's actually about 20x faster on my machine
+            (sw2.Elapsed).Should().BeLessThan(sw.Elapsed);
```

Comment:
These kind of tests can make the build server unstable.  When there is a drop of machine capacity in the second part of this unit test, the test could fail.

Reply:
That's why I kept a huge margin. I'll see if I can turn this into benchmarks instead.